drivers/display/display_renesas_lcdc.c

```diff
@@ -0,0 +1,595 @@
+{
+	const struct display_smartbond_config *config = dev->config;
+
+	memset(capabilities, 0, sizeof(*capabilities));
+
+	/*
+	 * Multiple color formats should be supported by LCDC. Currently, RGB56 and ARGB888
+	 * exposed by display API are supported. In the future we should consider supporting
+	 * more color formats which should require changes in LVGL porting.
+	 * Here, only one color format should be supported as the frame buffer is accessed
+	 * directly by LCDC and is allocated statically during device initialization. The color
+	 * format is defined based on the pixel-format property dictated by lcd-controller
+	 * bindings.
+	 */
+	capabilities->supported_pixel_formats = DT_INST_PROP(0, pixel_format);
+	capabilities->current_orientation = DISPLAY_ORIENTATION_NORMAL;
+	capabilities->current_pixel_format = DT_INST_PROP(0, pixel_format);
+	capabilities->x_resolution = config->x_res;
+	capabilities->y_resolution = config->y_res;
+}
+
+static int display_smartbond_read(const struct device *dev,
+			const uint16_t x, const uint16_t y,
+			const struct display_buffer_descriptor *desc,
+			void *buf)
+{
+	struct display_smartbond_data *data = dev->data;
+	uint8_t *dst = buf;
+	const uint8_t *src = data->buffer;
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	/* pointer to upper left pixel of the rectangle */
+	src += (x * data->current_pixel_size);
+	src += (y * data->layer.stride);
+
+	data->dma_block_cfg.block_size = desc->width * data->current_pixel_size;
+	/*
+	 * Source and destination base address is word aligned.
+	 * Data size should be selected based on color depth as
+	 * cursor is shifted multiple of pixel color depth.
+	 */
+	data->dma_cfg.source_data_size = data->dma_cfg.dest_data_size =
+			!(data->current_pixel_size & 3) ? 4 :
+			!(data->current_pixel_size & 1) ? 2 : 1;
+
+	data->dma_cfg.dest_burst_length = data->dma_cfg.source_burst_length =
+		!((data->dma_block_cfg.block_size / data->dma_cfg.source_data_size) & 7) ? 8 :
+		!((data->dma_block_cfg.block_size / data->dma_cfg.source_data_size) & 3) ? 4 : 1;
+
+	for (int row = 0; row < desc->height; row++) {
+
+		data->dma_block_cfg.dest_address = (uint32_t)dst;
+		data->dma_block_cfg.source_address = (uint32_t)src;
+
+		if (dma_config(data->dma, data->dma_channel, &data->dma_cfg)) {
+			LOG_ERR("Could not configure DMA");
+			k_sem_give(&data->device_sem);
+			return -EIO;
+		}
+
+		if (dma_start(data->dma, data->dma_channel)) {
+			LOG_ERR("Could not start DMA");
+			k_sem_give(&data->device_sem);
+			return -EIO;
+		}
+
+		k_sem_take(&data->dma_sync_sem, K_FOREVER);
+
+		src += data->layer.stride;
+		dst += (desc->pitch * data->current_pixel_size);
+	}
+
+	if (dma_stop(data->dma, data->dma_channel)) {
+		LOG_WRN("Could not stop DMA");
+	}
+
+	k_sem_give(&data->device_sem);
+
+	return 0;
+}
+
+static int display_smartbond_write(const struct device *dev,
+				const uint16_t x, const uint16_t y,
+				const struct display_buffer_descriptor *desc,
+				const void *buf)
+{
```

Comment:
I'm not familiar with this LCD controller, but something that has been implemented for the ELCDIF (which is designed to stream a buffer in memory directly to the display) is to switch to the user provided buffer directly if the buffer is the same size as the output display.

https://github.com/zephyrproject-rtos/zephyr/blob/eaf140e7ff063c2100f8c7dc4402cf6f7ba44506/drivers/display/display_mcux_elcdif.c#L94

I'm not sure if this is something your controller supports, but it might be worth investigating- it improved performance quite a bit using LVGL with the ELCDIF

Reply:
> What is also not clear to me is, if LCDC employs application buffer directly then any rending performed during frame update shouldn't result in visual artifacts?

The trick here would be to queue the application buffer somehow, and not load it into the LCDC hardware until the last frame was done streaming. Again, I am *not* familiar with the hardware here- so this should be taken as an idea, nothing more. No need to implement it if it is not possible or not worth looking into

Reply:
I see - such an architectural scheme should not work for us. However, I will spend some time on the underlying implementation :)